Zend/zend_language_parser.y

```diff
@@ -1273,7 +1276,11 @@ encaps_var:
 	|	T_DOLLAR_OPEN_CURLY_BRACES T_STRING_VARNAME '[' expr ']' '}'
 			{ $$ = zend_ast_create(ZEND_AST_DIM,
 			      zend_ast_create(ZEND_AST_VAR, $2), $4); }
-	|	T_CURLY_OPEN variable '}' { $$ = $2; }
+	|	T_CURLY_OPEN variable '}'
+			{
+				$$ = $2;
+				zend_error(E_DEPRECATED, "Array and string offset access syntax with curly braces is deprecated");
```

Comment:
This is unrelated to array access syntax.

Reply:
Yeah my mistake, would adding something comparable to:
```
	|	T_VARIABLE '{' encaps_var_offset '}'
	        { $$ = zend_ast_create(ZEND_AST_DIM,
                  zend_ast_create(ZEND_AST_VAR, $1), $3);
			      zend_error(E_DEPRECATED, "Array and string offset access syntax with curly braces is deprecated"); }
```
in ``encaps_var:`` do the trick?
(after also moving the direct calls to ``zend_error()`` in zend_compile)

Reply:
Why would this be needed? The tests seem to be passing with only the change under `dereferencable '{' expr '}'`.

Reply:
Maybe my local build is messed up but it seemed it had problems to throw a deprecation for a case (null coalesce maybe?) but if it works just fine without that's good. Now the question is why doesn't that work if you throw the deprecation from the compiler. 🤔

Reply:
You mean when adding the `ZEND_ALTERNATIVE_ARRAY_SYNTAX` AST attribute? It does work, except that this attribute gets overridden when using the null coalescing operator. I was able to work around this by treating the attribute as a bitmask, though.